HACKING.rst

```diff
@@ -86,18 +86,129 @@ You can also see all the environments by simply running ``tox list``
 Running ``tox run -m format`` and ``tox run -m lint`` before committing code is
 recommended.
 
-Code Style
-==========
-
 Commits
-#######
+-------
+
+Commit messages are based on the `conventional commit`_ style::
+
+  <type>(<optional scope>): <description>
+
+  <optional body>
+
+  <optional footer>
+
+The commit is divided into three sections: a header, body, and footer.
+
+Header
+======
+
+The header is required and consists of three subsections: a type,
+optional scope, and description. The header must be 72 characters or less.
+
+Types
+#####
+
+``build``
+"""""""""
+
+Commits that affect the build of an application or library.
+
+This includes dependency updates, which should use the ``deps`` scope
+(``build(deps):``).
+
+``chore``
+"""""""""
+
+Miscellaneous commits that don't fit into any other type.
+
+Examples include:
+
+* development dependency updates
+* edits to a comment or docstring
+* type changes
+* accommodating a developer-facing deprecation warning
+* many *small* fixes for an existing PR
+
+``ci``
+""""""
+
+Commits that affect the CI/CD pipeline.
+
+``docs``
+""""""""
+
+Commits that affect the contents of the documentation.
+
+Changes to how documentation is built should use the ``build`` type.
+
+Changes to how the documentation is built in the CI/CD pipeline should use
+the ``ci`` type.
+
+``feat``
+""""""""
+
+Commits that add a new feature for the user.
+
+``fix``
+"""""""
 
-Format your commits following the conventional_ commit style.
+Commits that fix a bug or regression.
 
-Optionally, use the parens to scope to a particular component where
-applicable.
+``perf``
+""""""""
 
-See below for some examples of commit headings::
+Commits that improve performance without changing the API or external behavior.
+
+``refactor``
+"""""""""""""
+
+Commits that refactor code.
+
+Using `Martin Fowler's definition`_, refactor means "*a change made
+to the internal structure of software to make it easier to understand and
+cheaper to modify without changing its observable behavior.*"
+
+``style``
+""""""""""
+
+Commits that change the syntax or format of the code. The meaning of the
+code should not change.
+
+This includes automatic changes from tools like ``black`` and ``ruff format``.
+
+
+``test``
+""""""""
+
+Commits that improve, add, or remove tests.
```

Comment:
What's the correct way to mark a refactor of tests?

Reply:
You could write it as `test(project): reorganize tests` or `refactor(test): reorganize project tests`, but I'm not sure which is most correct.

`docs` and `test` are challenging because they are `types` but can also be a `scope`.

Do you have an opinion?  We should address this scenario one way or the other.

Reply:
I do feel`refactor(test)` more natural

Reply:
I agree that `refactor(test)` feels more natural. I also think we might even be able to give general guidance on priority for types. In descending priority I'd say the types are roughly:

ci, build, feat, fix, refactor, style, test, docs, chore

Obviously it's preferable to break up most things that can be broken up to prevent type conflicts (e.g. stick the code and the docs in separate PRs if at all possible), but sometimes this isn't possible. This does come with qualifications though:

* `feat` and `fix` PRs should always include tests, and many `refactor` PRs will include tests as well.
* `style` PRs can cover any files (code, docs, config, etc.) as long as the changes are purely aesthetic and not functional.
* Many higher priority types might include changes to CI and build too, but we don't need to call this out.
* `ci` and `build` types are invalid if they contain anything other than ci or build changes. CI is higher priority than build because it's a narrower scope - many `build` PRs will contain CI changes, but no `ci` PRs will touch the build system outside of CI. Likewise, if something is tagged with `build` it will almost certainly not contain a change in the source directory or the tests directory.

Reply:
In fact if we like that order, maybe we should reorder the types according to that, giving the general guidance of "if you're unsure what to choose, choose the first-listed type that correctly describes your commit"

Reply:
Sounds good.  I think you missed `perf` by accident, so I placed it before `refactor` 

Some references call `perf` a subcategory of refactor which can be true but I don't want to write that down, since I gave a very strict definition of `refactor` here.

Reply:
I did, and I agree with your location. I'm mixed on calling `perf` a subcategory of `refactor`, since I view `refactor` as something that's for developers, whereas `perf` is "for the machine."